### PR TITLE
UMA07 - Large Relay Detection bot

### PR DESCRIPTION
**Pull request by MakC-Ukr**
# Large Relay Detection bot

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Bot
- [ ] Fix a bot error
- [ ] Update a bot logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Bot PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the bot properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The bot README is updated
- [x] The bot README sufficiently describes what the bot does
- [x] All the alerts that the bot emits are listed in the README
- [x] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Bot contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [x] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the bot is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The bot follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Bot can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

If there is some useful information or important note for reviewing/understanding this bot please describe it here.

### CODE REVIEW

File: UMA-Bots/Large-Relays/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+In order to change the token amount thresholds, please add/remove the token addresses in the `./src/chainThresholds.ts` file or change the thresholds amounts for specific chains. Please note that the currently added tokens and corresponding amounts are tentative.
+
+## Test Data
+
+The bot behaviour can be verified with the following transactions by running `npm run tx <TX_HASH>`:
+
+### Ethereum Mainnet
+- [0x51fa8f3cabfe44033bfd4729a60eb6d8c57c54a3097463207e33e218b9a91d35](https://etherscan.io/tx/0x51fa8f3cabfe44033bfd4729a60eb6d8c57c54a3097463207e33e218b9a91d35) (1 finding - `FilledRelay` was emitted 1 time with an amount of 3.87 WETH relayed)
+- [0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9](https://etherscan.io/tx/0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9) (1 findings - `FilledRelay` was emitted 1 time with an amount of 120 USDC relayed)
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/README.md:38]:
```suggestion
- [0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9](https://etherscan.io/tx/0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9) (1 finding - `FilledRelay` was emitted 1 time with an amount of 120 USDC relayed)
```

File: UMA-Bots/Large-Relays/src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      destinationChainId,
+      depositor,
+      recipient,
+      isSlowRelay,
+    },
+  });
+}
+
+// @Review Since the PoC addresses don't really need to be configurable by the client,
+// is it better to add the Goerli addresses here or in chainThresholds.ts ?
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/utils.ts:44]: Surely not a big deal, I personally would still move it to the chainThresholds.

File: UMA-Bots/Large-Relays/src/agent.spec.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import { TestTransactionEvent } from "forta-agent-tools/lib/test";
+import { FILLED_RELAY_EVENT } from "./utils";
+import { provideHandleTransaction } from "./agent";
+import { createAddress, NetworkManager } from "forta-agent-tools";
+import { NetworkDataInterface } from "./network";
+import { BigNumber } from "ethers";
+// @Review I believe that here BigNumber is being imported from "ethers" that is exported from "forta-agent".
+// Since we dont have to install the ethers.js as a separate library here (not there in package.json), I assume this import is fine.
+// Not 100% sure though. In case we are actually using ethers.js as a separate library (which doesn't make sense since "forta-agent" includes it),
+// we can import ethers from "forta-agent" and then call `ethers.BigNumber.from()`
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/agent.spec.ts:11]: It's fine as it is 👍

File: UMA-Bots/Large-Relays/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+describe("Large relay detection bot test suite", () => {
+  let handleTransaction: HandleTransaction = provideHandleTransaction(FILLED_RELAY_EVENT, networkManagerTest);
+
+  it("returns empty findings if there is no event emitted", async () => {
+    const txEvent: TransactionEvent = new TestTransactionEvent();
+    const findings = await handleTransaction(txEvent);
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("doesn't return a finding if a large amount of specified tokens are bridged via a contract other than the whitelisted SpokePool", async () => {
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/agent.spec.ts:103]: Can you explain the "whitelisted" ?

File: UMA-Bots/Large-Relays/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        FILLED_RELAY_EVENT,
+        TEST_SPOKEPOOL_ADDR,
+        passParams(MOCK_THRESHOLD_ERC20.toString(), MOCK_ERC20_ADDR)
+      )
+      .addEventLog(
+        FILLED_RELAY_EVENT,
+        TEST_SPOKEPOOL_ADDR,
+        passParams(MOCK_THRESHOLD_ERC20.toString(), MOCK_ERC20_ADDR_2)
+      )
+      .addEventLog(RANDOM_EVENT_ABI, TEST_SPOKEPOOL_ADDR, [RANDOM_ADDRESSES[0], "120"]) // Non-whitelisted SpokePool events shall be ignored
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/agent.spec.ts:178]: This "non-whitelisted" basically means "other", correct?

File: UMA-Bots/Large-Relays/src/agent.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    const findings: Finding[] = [];
+    const filledRelayEventTxns = txEvent.filterLog(filledRelayEvent, networkManager.get("spokePoolAddr"));
+
+    filledRelayEventTxns.forEach((filledRelayEvent) => {
+      const { amount, originChainId, destinationChainId, depositor, recipient, isSlowRelay, destinationToken } =
+        filledRelayEvent.args;
+
+      if (
+        Object.keys(networkManager.get("tokenThresholds")).includes(destinationToken) &&
+        BigNumber.from(amount.toString()).gte(BigNumber.from(networkManager.get("tokenThresholds")[destinationToken]))
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/agent.ts:32]:
```suggestion
        amount.gte(BigNumber.from(networkManager.get("tokenThresholds")[destinationToken]))
```

**Review by Vxatz — CHANGES_REQUESTED** [review]
Good job once again 🎉

File: UMA-Bots/Large-Relays/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
+# Large Relays detection bot
+
+## Description
+
+This bot monitors for large relays, such as those over 250 ETH, 1,000,000 USDC, 20 WBTC, and so on made with [Across v2 bridge](https://across.to/) - a multichain bridge built on the [UMA protocol](https://umaproject.org/) as its source of on-chain data validation. For more details refer [here](https://discourse.umaproject.org/t/forta-monitors-across-v2-request-for-proposals/1569).
```
[the remark this answers is not in the document]
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/README.md:5]: Yep, done !)

File: UMA-Bots/Large-Relays/src/utils.ts (hunk 2 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/src/utils.ts:44]: Sure thing. thanks for confirming

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 4 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:103]: By "whitelisted" I meant the Across's SpokePool address. We can rename it to "official" if it makes thing a bit clearer. Wdyt ?

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 5 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:178]: Yep

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 4 above)
Comment (Vxatz) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:99]: Yep, I think "official" is better

File: UMA-Bots/Large-Relays/src/network.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  ARBITRUM_SPOKEPOOL,
+  Dictionary,
+  GOERLI_POC_SPOKEPOOL_ADDRESS,
+  MAINNET_SPOKEPOOL,
+  OPTIMISM_SPOKEPOOL,
+  POLYGON_SPOKEPOOL,
+} from "./utils";
+
+export interface NetworkDataInterface {
+  tokenThresholds: Dictionary<string>;
```
Comment (Vxatz) [inline UMA-Bots/Large-Relays/src/network.ts:20]:
```suggestion
  tokenThresholds: Record<string, string>;
```

Could use this

File: UMA-Bots/Large-Relays/README.md (hunk 7 above)
Comment (Vxatz) [inline UMA-Bots/Large-Relays/README.md:5]: It's better to either avoid mentioning the amounts (250, 1m, 20) here or change the thresholds to these

File: UMA-Bots/Large-Relays/README.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+In order to change the token amount thresholds, please add/remove the token addresses in the `./src/chainThresholds.ts` file or change the thresholds amounts for specific chains. Please note that the currently added tokens and corresponding amounts are tentative.
+
+## Test Data
+
+The bot behaviour can be verified with the following transactions by running `npm run tx <TX_HASH>`:
+
+### Ethereum Mainnet
+- [0x51fa8f3cabfe44033bfd4729a60eb6d8c57c54a3097463207e33e218b9a91d35](https://etherscan.io/tx/0x51fa8f3cabfe44033bfd4729a60eb6d8c57c54a3097463207e33e218b9a91d35) (1 finding - `FilledRelay` was emitted 1 time with an amount of 3.87 WETH relayed)
+- [0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9](https://etherscan.io/tx/0x396c794b8a41e6e365a0fc52235739c6e82751b977d3f803d622c9463713e1d9) (1 finding - `FilledRelay` was emitted 1 time with an amount of 120 USDC relayed)
```
Comment (0xtaf) [inline UMA-Bots/Large-Relays/README.md:38]: this gives 2 findings actually

File: UMA-Bots/Large-Relays/src/network.ts (hunk 8 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/src/network.ts:20]: sure, thanks

File: UMA-Bots/Large-Relays/README.md (hunk 7 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/README.md:5]: Sure, I will replace the amounts with `configurable thresholds`

File: UMA-Bots/Large-Relays/README.md (hunk 9 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/README.md:38]: Thanks for catching this. Edited

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 4 above)
Comment (0xtaf) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:99]: just reminding this @MakC-Ukr

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 5 above)
Comment (0xtaf) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:174]: can be unofficial? according to the convo above

File: UMA-Bots/Large-Relays/README.md (hunk 9 above)
Comment (0xtaf) [reply UMA-Bots/Large-Relays/README.md]: should be 2 finding`s` 😄

File: UMA-Bots/Large-Relays/src/agent.spec.ts (hunk 5 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/src/agent.spec.ts:174]: I just rephrased it. Please check if it is more readable now 😃

File: UMA-Bots/Large-Relays/README.md (hunk 9 above)
Comment (MakC-Ukr) [reply UMA-Bots/Large-Relays/README.md]: Just Changed 😸

**Review by Vxatz — APPROVED** [review]

**Review by 0xtaf — APPROVED** [review]
another great job 👏

**Comment by MakC-Ukr** [comment]
Ready for merging :)  @RCantu92

**Pull request merged**